leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClientBuilder.java

```diff
@@ -57,13 +59,17 @@
  * Helper class to build and configure a Californium based Leshan Lightweight M2M client.
  */
 public class LeshanClientBuilder {
+    static {
+        CoapConfig.register();
+        DtlsConfig.register();
```

Comment:
Before the redesign, it was even "statically compiled in" :-).
I think, if you try to play with, you will find out, that for reading the files, it's very important to have all definitions in place.

Reply:
> for reading the files, it's very important to have all definitions in place.

I'm not sure to get it. As to read I need to call this `register` anyway ? 
So finally I need to be aware that I need to register some config definition.

So what the difference with the dynamic one ? 

Imagine that each config provides a `DefinitionsProvider`
Then you can just do  something like : 
```java
Configuration networkConfig = new Configuration(CoapConfig.getDefinitions(), DtlsConfig.getDefinitions()); 
Configuration networkConfig = Configuration.createWithFile("myfile", "myheader", CoapConfig.getDefinitions(), DtlsConfig.getDefinitions());
```
If you want not to register it several time you can imagine a kind of Factory where you register the definition and you use it later. (if someone want it static it the create a static ref)

(I don't really need this but if you want to be able to load file without register definitions before, maybe there is a way storing full class name of `DefinitionsProvider` in the config so being able to load it dynamically ?)

Reply:
> As to read I need to call this register anyway ?

Yes, but only once per application or per application-type (e.g. `cf-cli` for clients).

Reply:
Generally, I don't want to spend my time in testing a "more dynamic" solution (years ago, converting the static timer in californium, that was about two weeks of searching in rare failing unit tests, to sort out the "hidden" consequences in too many dirty unit tests.).
 
I also don't want to block your idea. So, what's about having the current `static` definitions as the variant I use in Californium. And an additional `Configuration` with the provided definitions for your purpose (and you test and maintain it)?

Reply:
Maybe, even better, you may add your solution additionally to the `Configuration`?

Reply:
About the feature itself I don't get your point : 

> Yes, but only once per application or per application-type (e.g. cf-cli for clients).

Generally the application need to initialize configuration once anyway.
When you need to create more you maybe also need to load different definition so dynamic is better.
If you want to create several config with same definitions registered you create a Factory that you can reuse. (it perfectly match the cf-cli  and leshan use-case).

Using static state is well known as bad practice, I'm sure like all bad practice sometime it make sense to use it anyway but here I can not see any benefits :thinking:

Reply:
I prepared the [Californium PR1701](https://github.com/eclipse/californium/pull/1701). That's at least a first version to start.

Reply:
> Generally the application need to initialize configuration once anyway.

Yes, and now you can provide the modules you want on your responsibility.

Reply:
> Using static state is well known as bad practice

That seems to depend on the use-case.
And the time, people want to spend in other solutions.

Reply:
About the way you handle this proposition I'm a bit confused : 

> you may add your solution additionally to the Configuration

Eventually I could try to propose a PR. (Even if all of this are also not planned to me)

> (and you test and maintain it)?

I'm not sure to understand concretely how this should looks like ?

> Generally, I don't want to spend my time in testing a "more dynamic" solution  (years ago, converting the static timer in californium, that was about two weeks of searching in rare failing unit tests, to sort out the "hidden" consequences in too many dirty unit tests.).

This doesn't make me comfortable to contribute. 

I also spend my time testing and fixing contribution from other people. I don't ask to Leshan contributors to maintain their code contribution for ever. I also spend my time searching issue because of Californium bugs or changes that I didn't request. I feel this is how open source works. 

I'm not sure that  _"I don't want to waste my time with your idea, so eventually do you stuff in the project and handle it on your own"_ is something which could work (especially for a core feature)

> So, what's about having the current static definitions as the variant I use in Californium. And an additional Configuration with the provided definitions for your purpose

If we put the dynamic way in cf code base, It would be very hard to me to understand why californium code/test will not use this dynamic way. But maybe I missed something obvious. 


> And the time, people want to spend in other solutions.

Of course it's time consuming to first write it statically then change it to a dynamic way.

> I prepared the Californium PR1701. That's at least a first version to start.

I'm still confused, will you go further in that way or do you still expected a PR from me or ?

Reply:
> I prepared the Californium PR1701. That's at least a first version to start.

Thx to be a more bit open to the idea.

Reply:
FMPOV, there is a solution in Californium. You want something different, which in my opinion doesn't have the benefit, which makes me spending too much time in it. what I don't want, is to remove the static stuff, it's more about adding a way you want to use in Leshan.

Reply:
Maybe we should get other opinions.

As I said I try to spend some time to provide code but obviously saying OK you can contribute but I don't want your code to be used in californium is not so engaging ...

Reply:
My point is: for me, the current solution in Californium works! It replaces an old solution and improves the function. It's now typed, offers modules and a documented property file.

If you want a solution, where the modules are specified when creating the Configuration in Leshan, OK go for it, additionally. For that some adaptions in Californium are required and so I did a short check, if that could work at all. But I didn't check, if that is all, which need to be adapted nor if that "fully works". I guess, that not too many additional changes are required for Californium.

For me, the main work left is in leshan using and testing that approach. I considered your motivation was to be able to use such an approach in leshan.

Reply:
> I considered your motivation was to be able to use such an approach in leshan.

I don't always propose changes for californium as a Leshan user but sometime also as a californium committer but I feel that you consider my point mainly when I have my Leshan user hat so finally I'm less and less implied as californium committer and just as californium user.